Clarify purposes and processes in the Sentences documentation (#3804)
common-voice/common-voice: docs/SENTENCES.md

As Common Voice is a read dataset, sentences are our currency. You can help by adding new sentences to our dataset for other contributors to read, helping with bulk sentence extractions, or reporting problematic sentences.

## In a few words...

> 📝 Sentence collector is made for everyday users to have a webpage (the [Sentence Collector webpage](https://commonvoice.mozilla.org/sentence-collector/)) allowing them to load some sentences 'directly' from the Common Voice webpage. It has it's own repo in [common-voice/sentence-collector](https://github.com/common-voice/sentence-collector). **This is a good place to start** for newcomers to this project.

Author: here is my proposal : 

📝 [Sentence collector](https://github.com/common-voice/sentence-collector) is a tool for contributors to upload public domain sentences through a webpage (the [Sentence Collector webpage](https://commonvoice.mozilla.org/sentence-collector/)) which then can get reviewed and are exported to the [Common Voice database](https://github.com/common-voice/common-voice/tree/main/server/data). Once imported into the Common Voice website, they will show up for contributors to read out aloud. This is a good place to start for newcomers to this project.

> 📝 Sentence collector is made for everyday users to have a webpage (the [Sentence Collector webpage](https://commonvoice.mozilla.org/sentence-collector/)) allowing them to load some sentences 'directly' from the Common Voice webpage. It has it's own repo in [common-voice/sentence-collector](https://github.com/common-voice/sentence-collector). **This is a good place to start** for newcomers to this project.

Author: And modified as requested in https://github.com/common-voice/common-voice/pull/3804#discussion_r974700174

> 📝 Sentence collector is made for everyday users to have a webpage (the [Sentence Collector webpage](https://commonvoice.mozilla.org/sentence-collector/)) allowing them to load some sentences 'directly' from the Common Voice webpage. It has it's own repo in [common-voice/sentence-collector](https://github.com/common-voice/sentence-collector). **This is a good place to start** for newcomers to this project.

Reviewer: Sounds good, however I would prefer

> ... through a [webpage](https://commonvoice.mozilla.org/sentence-collector/) which then...

so we don't need to repeat words.

📘 For power users who want's to [bulk](https://en.wikipedia.org/wiki/Bulk_insert) uploads of thousands of sentences, like for books, Sentence Collector is not the best tool, and they _should definitely_ check out below the [Bulk Submission](https://github.com/common-voice/common-voice/blob/main/docs/SENTENCES.md#bulk-submission) guidelines for this use case 😉. There is no dedicated repo.

🖥️ Finally, for automatic extraction of data sources, the Sentence Extractor repo ([common-voice/cv-sentence-extractor](https://github.com/Common-Voice/cv-sentence-extractor)) is dedicated for it. It can be Wikipedia, Wikisource, raw file... 

## Sentence Collector

> The [Sentence Collector](https://commonvoice.mozilla.org/sentence-collector/) webpage is a tool for crowdsourcing sentences for Common Voice. You can either:

Author: As a confused newbie I was, sincerely, yes it is 🤣 
I made days to catch that the sentence collector repo was the webpage. It's obvious when you know... But you have to know first. Better to be written down !

> The [Sentence Collector](https://commonvoice.mozilla.org/sentence-collector/) webpage is a tool for crowdsourcing sentences for Common Voice. You can either:

Author: modified 👍

> - Add sentences for your language (there is a few cleanup and validation routines under the hood)

Author: Yes... I was thinking that we could say something like "(go to the repository for details)"

> - Add sentences for your language (there is a few cleanup and validation routines under the hood)

Author: Here is my proposal :

## Sentence Collector

The [Sentence Collector](https://commonvoice.mozilla.org/sentence-collector/) is a website for crowdsourcing sentences for Common Voice. You can either:

- Add sentences for your language 
- Validate sentences that other contributors have added

Each sentence requires at least two upvotes from human validation to be considered valid.

Every week, validated sentences from the Sentence Collector will be exported and added to the Common Voice repository, and will be available at the next release of the Common Voice website. 

For more detailled explanations, see the [ReadMe file](https://github.com/common-voice/sentence-collector/blob/main/README.md) of the Sentence Collector.

As Common Voice is a read dataset, sentences are our currency. You can help by adding new sentences to our dataset for other contributors to read, helping with bulk sentence extractions, or reporting problematic sentences.

## In a few words...

> 📝 [Sentence collector](https://github.com/common-voice/sentence-collector) is a tool for contributors to upload public domain sentences through a website (the [Sentence Collector website](https://commonvoice.mozilla.org/sentence-collector/)) which then can get reviewed and are exported to the [Common Voice database](https://github.com/common-voice/common-voice/tree/main/server/data). Once imported into the Common Voice website, they will show up for contributors to read out aloud. This is a good place to start for newcomers to this project.

Reviewer: ```suggestion
📝 [Sentence collector](https://github.com/common-voice/sentence-collector) is a tool for contributors to upload public domain sentences through a [website](https://commonvoice.mozilla.org/sentence-collector/) which then can get reviewed and are exported to the [Common Voice database](https://github.com/common-voice/common-voice/tree/main/server/data). Once imported into the Common Voice website, they will show up for contributors to read out aloud. This is a good place to start for newcomers to this project.
```

📘 Contributors who want to [bulk upload](https://en.wikipedia.org/wiki/Bulk_insert) thousands of sentences, like for books, should check out the [Bulk Submission](https://github.com/common-voice/common-voice/blob/main/docs/SENTENCES.md#bulk-submission) guidelines below. There is no dedicated repository for this.

🖥️ For automatic extraction of data sources, the [Sentence Extractor](https://github.com/Common-Voice/cv-sentence-extractor) is dedicated for extracting from sources such as Wikipedia, Wikisource or raw files.

## Sentence Collector

The [Sentence Collector](https://commonvoice.mozilla.org/sentence-collector/) is a website for crowdsourcing sentences for Common Voice. You can either:

- Add sentences for your language 
- Validate sentences that other contributors have added

Each sentence requires at least two upvotes from human validation to be considered valid.

Every week, validated sentences from the Sentence Collector will be exported and added to the Common Voice repository, and will be available at the next release of the Common Voice website. 

> For more detailled explanations, see the [ReadMe file](https://github.com/common-voice/sentence-collector/blob/main/README.md) of the Sentence Collector.

Reviewer: minor: typo in "detailled" -> "detailed"

Reviewer: ```suggestion
For more detailed explanations, see the [README file](https://github.com/common-voice/sentence-collector/blob/main/README.md) of the Sentence Collector.
```

## Automatic extraction

> The Sentence Extractor is a tool that can scrape public domain data sources for sentences. There are multiple sources integrated into the Sentence Extractor, such as Wikipedia and Wikisource. Please [see this post](https://discourse.mozilla.org/t/sentence-extractor-current-status-and-workflow-summary/62332) for detailed guidance on how to use the Sentence Extractor.

Reviewer: Haven't noticed this before, but probably wouldn't be bad to link to it here. Even if it's already linked further above.

```suggestion
The [Sentence Extractor](https://github.com/Common-Voice/cv-sentence-extractor) is a tool that can scrape public domain data sources for sentences. There are multiple sources integrated into the Sentence Extractor, such as Wikipedia and Wikisource. Please [see this post](https://discourse.mozilla.org/t/sentence-extractor-current-status-and-workflow-summary/62332) for detailed guidance on how to use the Sentence Extractor.
```

Once the review is complete, submit a pull request with the # of sentences submitted, a link to the manual QA results, and the % error rate. Here's [an example PR](https://github.com/mozilla/common-voice/pull/2873). Please make sure the sentences are in a plain `.txt` file with one sentence per line.

## Rules for QA that applies (or not) to the different imputs and outputs

> To help you, for exemple with the 'not supervised, do your own QA' bulk submission, you _may_ find interesting to have a look at : 

Reviewer: While this may be helpful, it is not a rule for QA of bulk submissions, so I'm a bit hesitant in listing this here as is. @ftyers do you have thoughts on this?

Author: Indeed, QA is requested... But not "how to do it"! you're right. 

Here is the reworked version : 

## QA that applies (or not) to the different imputs and outputs
As a reminder, here is listed the different automated QA applied, depending of the process.
* [cleanup](https://github.com/common-voice/sentence-collector/tree/main/server/lib/cleanup) and [validation](https://github.com/common-voice/sentence-collector/tree/main/server/lib/validation) rules for Sentence Collector, 
* [language](https://github.com/common-voice/cv-sentence-extractor#using-language-rules) rules for Sentence Extractor, 
* [preprocessor](https://github.com/common-voice/CorporaCreator/tree/master/src/corporacreator/preprocessors) for CorporaCreator ([CorporaCreator](https://github.com/common-voice/CorporaCreator) is the Common Voice extractor that, linking text and recorded voice samples, make them available for [SST](https://en.wikipedia.org/wiki/Speech_recognition) engines' training).

To help you, for exemple with the 'not supervised, do your own QA' bulk submission, you may find interesting to have a look at it.

* [cleanup](https://github.com/common-voice/sentence-collector/tree/main/server/lib/cleanup) and [validation](https://github.com/common-voice/sentence-collector/tree/main/server/lib/validation) rules for sentence collector, and/or 
* [language](https://github.com/common-voice/cv-sentence-extractor#using-language-rules) rules for extractor, or even 
* [preprocessor](https://github.com/common-voice/CorporaCreator/tree/master/src/corporacreator/preprocessors) for CorporaCreator ([CorporaCreator](https://github.com/common-voice/CorporaCreator) is the Common Voice extractor that, linking text and recorded voice samples, make them available for [SST](https://en.wikipedia.org/wiki/Speech_recognition) engines' training).

 
# Correcting existing data

> As you can see, some methods doesn't go through automated cleanup/validation/rules, and they are not unified ([...not yet](https://discourse.mozilla.org/t/sentence-collector-cleanup-before-export-vs-cleanup-on-upload/105411/15)). Thus, there is a process to remove old data that might need to be discarded.

Reviewer: ```suggestion
Some methods don't go through automated cleanup/validation/rules, and they are not unified ([...not yet](https://discourse.mozilla.org/t/sentence-collector-cleanup-before-export-vs-cleanup-on-upload/105411/15)). Thus, there is a process to remove old data that might need to be discarded.
```

.. additionally, I do not think this is the right link to link. If anything, this should point to a dedicated discussion thread around this topic, or a GitHub issue. Linking just that comment does not provide more info that "not yet" states.

Additionally, this is not the only reason why a sentence would be removed. I would like to keep this more general here.

Author: I don't want to start a fire that I won't be able to handle, but I don't understand why cleanup and validation are not unified. ...Is it by choice / design, or only by "legacy" ? 

Shall we open an issue saying "We should unify cleanup and validation between sentence collector, sentence extractor and corpora creator" ?

Reviewer: > I don't want to start a fire that I won't be able to handle, but I don't understand why cleanup and validation are not unified. ...Is it by choice / design, or only by "legacy" ?

Out of those choices, I'd say legacy applies best here. In the end these tools also use different technologies, however that would be solve-able in the end. There are also slight differences in purposes, however that could probably also be dealt with, though I have to admit that I don't know the exact workings of the CorporaCreator.

Yes, something to track a possible "one rule file fits all" solution would be great, and that, in my opinion, could then also be linked here. As this spans multiple repos, maybe we should start with a [Discourse](https://discourse.mozilla.org/c/voice/239) post, and then go from there and once we agree on something create issues in the appropriate repositories.

Author: I'm starting with "let's have an overview", then I'll follow by "now we can see that we do everywhere some cleanup and validation, shall we have a unified verification and cleanup preprocessor, as for all inputs, and before output/corpora creator"

I think with solve this conversation...

https://discourse.mozilla.org/t/trying-to-create-a-common-voice-overview-for-newbies-svg-drawin-style/105771

## Flagging (and removing) problematic sentences already in the Common Voice database

> If you notice sentences that need to be deleted 🚩, first check what the source of the sentence is. 

Reviewer: ```suggestion
If you notice sentences that need to be deleted, first check what the source of the sentence is. 
```

> If the file that the sentence is located in is called `sentence-collector.txt` (localised [here](https://github.com/common-voice/common-voice/blob/main/server/data/en/sentence-collector.txt) for english, or [here](https://github.com/common-voice/common-voice/blob/main/server/data/fr/sentence-collector.txt) for french, etc.), that means it was automatically exported from Sentence Collector. In that case, please file an [issue](https://github.com/common-voice/sentence-collector/issues) on the [Sentence Collector repo](https://github.com/Common-Voice/sentence-collector/) with a plaintext file of all problematic sentences, listed one sentence per line. (Note : it is useless to correct directly the common voice database if it comes for sentence collector, because it will automatically be imported, again, by the sentence collector. You have to clean up sentence collector's database first)

Reviewer: ```suggestion
Search for the source of the sentence within the [data](https://github.com/common-voice/common-voice/blob/main/server/data) folder. Folders are split up by language. If the sentence is found in `sentence-collector.txt`, that means it was automatically exported from Sentence Collector. In that case, please file an [issue](https://github.com/common-voice/sentence-collector/issues) with a plaintext file of all problematic sentences, listed one sentence per line. Note that in this case removing it from the file through a Pull Request will not help, because it will automatically be added again with the next export.
```

I've rephrased it a bit so examples shouldn't be necessary. What do you think?

Author: yes :) 👍

Once the review is complete, submit a pull request with the # of sentences submitted, a link to the manual QA results, and the % error rate. Here's [an example PR](https://github.com/mozilla/common-voice/pull/2873). Please make sure the sentences are in a plain `.txt` file with one sentence per line.

> ## QA that applies (or not) to the different imputs and outputs

Reviewer: ```suggestion
## QA that applies (or not) to the different inputs and outputs
```

however I'm not sure this title is fully understandable. Maybe "Cleanup/Processing" might be better? Not really convinced of that either though.

> As a reminder, here is listed the different automated QA applied, depending of the process.

Reviewer: ```suggestion
Depending of the process different automated transformations applied:
```

> As a reminder, here is listed the different automated QA applied, depending of the process.

Author: ```suggestion
Depending of the process, different automated transformations are applied:
```

Reviewer: Oh, I just noticed a typo in my initial suggestion for this :'( should be "Depending **on** the process"

* [cleanup](https://github.com/common-voice/sentence-collector/tree/main/server/lib/cleanup) and [validation](https://github.com/common-voice/sentence-collector/tree/main/server/lib/validation) rules for Sentence Collector, 
* [language](https://github.com/common-voice/cv-sentence-extractor#using-language-rules) rules for Sentence Extractor, 

> * [preprocessor](https://github.com/common-voice/CorporaCreator/tree/master/src/corporacreator/preprocessors) for CorporaCreator ([CorporaCreator](https://github.com/common-voice/CorporaCreator) is the Common Voice extractor that, linking text and recorded voice samples, make them available for [SST](https://en.wikipedia.org/wiki/Speech_recognition) engines' training).

Reviewer: I'm not sure about this CorporaCreator description, so I'll let others review this.

Reviewer 2: CorporaCreator creates the default train, dev and test splits from validated recordings, taking a line of diversity rules and statistical correctness into account. These splits are used in training models for STT (Speech to Text).

PS: AFAIK, the preprocessors existing there are last minute correcting rules, as some (sometimes most) of the text-corpora are coming from bulk-submission PRs...

Reviewer 2: This might help about inner workings (the second part of the post):
https://discourse.mozilla.org/t/inadequate-documentation/105415/2

Author: Different clarifications are in progress, both on README files ([here](https://github.com/common-voice/common-voice/pull/3811)) and on an overview of process ([here](https://discourse.mozilla.org/t/trying-to-create-a-common-voice-overview-for-newbies-svg-drawin-style/105771))...

I think it resolve the conversation

> To help you, for exemple with the 'not supervised, do your own QA' bulk submission, you may find interesting to have a look at it. 

Reviewer: ```suggestion
To help you for example with the 'not supervised, do your own QA' bulk submission, you may find these resources interesting. 
```
